### ISSUE TITLE
Introduction of JsonTypeInfo <TValue> in one of the Serialize overloads breaks the code in .net 6

### ISSUE THREAD

**Issue by mayur-ekbote** [dotnet/efcore#25889]
I have EF Core conversion lambdas as follows:

`entity.Property(e => e.Value).HasConversion(v => JsonSerializer.Serialize(v,null), v => JsonSerializer.Deserialize(v, null));
`
where 'v' is a HashSet<string> . It worked in .net5

After moving to .net 6, I got this compilation error:

"The call is ambiguous between the following methods or properties: 'JsonSerializer.Serialize<TValue>(TValue, JsonSerializerOptions?)' and 'JsonSerializer.Serialize<TValue>(TValue, JsonTypeInfo<TValue>)'"

This is because a new overloaded Serialize was introduced as a performance update to JsonSerializer 
(via  https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-source-generator/ )?

However, I cannot keep null as a named parameter or omit the null as an optional parameter because it is not allowed in lambads ( CS0854, there is no documentation page for this error though: https://docs.microsoft.com/en-us/dotnet/csharp/misc/sorry-we-don-t-have-specifics-on-this-csharp-error?f1url=%3FappId%3Droslyn%26k%3Dk(CS0854))

Consequently the code needs to be changed to:

`entity.Property(e => e.Value).IsRequired().HasConversion(v => JsonSerializer.Serialize(v, new JsonSerializerOptions()), v => JsonSerializer.Deserialize<HashSet<string>>(v, new JsonSerializerOptions()));`

**Comment by roji** [comment]
@mayur-ekbote have you tried disambiguating by casting the null value, e.g. specifying `(JsonSerializerOptions)null`?

**Comment by mayur-ekbote** [comment]
No, didn't try it. I used `new JsonSerializerOptions() ` (and I don't see any difference in the output). My concern is only related to code-breaking.
I am not sure it is related to ef core. Neither of the assemblies pertains to ef core so I raised it in aspnetcore.

**Comment by ajcvickers** [comment]
@jeffhandley @danmoseley This is a breaking change in runtime, but I don't see it documented here: https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/obsolete-apis-with-custom-diagnostics Was this unintentional, or is it just waiting documentation?

**Comment by jeffhandley** [comment]
> @jeffhandley @danmoseley This is a breaking change in runtime, but I don't see it documented here: https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/obsolete-apis-with-custom-diagnostics Was this unintentional, or is it just waiting documentation?

Tagging @dotnet/area-system-text-json to get a response from that crew.

**Comment by layomia** [comment]
Ah, got you. I reckoned that because JSON serialization was pervasive it was worth creating breaking change documentation. Closing https://github.com/dotnet/docs/issues/25987.

**Comment by ericstj** [comment]
It doesn’t hurt to publish it, especially not if you feel it will be common to pass `null` for options,  but it’s not something we would typically consider breaking. Cc @terrajobst

**Comment by ajcvickers** [comment]
@ericstj It's worth noting that passing an expression tree to EF that uses this is _very_ common. It's even in our docs as an example. Many people will experience this break.

**Comment by eiriktsarpalis** [comment]
> However, I cannot keep null as a named parameter or omit the null as an optional parameter because it is not allowed in lambads ( CS0854, there is no documentation page for this error though: https://docs.microsoft.com/en-us/dotnet/csharp/misc/sorry-we-don-t-have-specifics-on-this-csharp-error?f1url=%3FappId%3Droslyn%26k%3Dk(CS0854))

I wasn't aware of this restriction, and I'm a bit surprised that it exists. Even if they cannot be represented in the underlying expression tree, named parameters could be used to provide disambiguate overload resolution. For context F# quotations do not have this restriction.

**Comment by bartonjs** [comment]
> We should be careful about adding overloads where it creates an ambiguous null parameter.

We generally are.  FDGv3 Appendix D (Breaking Changes)

> **D.10.2 Adding Alternative-Parameter Overloads for a Reference Type Parameter**
>
>When two different overloads accept reference types, the null (or default)
literal value is ambiguous across the two members. If the existing member
throws an ArgumentNullException when the varied parameter is null,
then the .NET BCL team considers it an acceptable breaking change to
introduce a compile-time break for those callers that pass null. However,
if the null value is accepted by the member, then overloading is only an
acceptable change if some other portion of the parameter list forces uniqueness
to remove the ambiguity of a null literal.

I think this slipped through the cracks because the existing overload's parameter used a nullable value type.  It has the exact same concerns, but we probably lumped it into "we don't care about breaking callers who use the bareword `default`", neglecting that the `null` literal also applies.

**Comment by terrajobst** [comment]
In general, every time we add an overload with a type that can accept `null` where previously there was no overload can cause these kinds of ambiguities. I don't think it's practical to avoid them in API design because it would severely restrict how we can evolve the APIs. The way they should be resolved is by casting `null` to the appropriate type. Personally, I don't believe it's worth a breaking change entry.
